packages/richtext-lexical/src/features/upload/server/nodes/UploadNode.tsx

```diff
@@ -17,28 +23,29 @@ import { $applyNodeReplacement } from 'lexical'
 import * as React from 'react'
 
 export type UploadData<TUploadExtraFieldsData extends JsonObject = JsonObject> = {
-  fields: TUploadExtraFieldsData
-  /** Every lexical node that has sub-fields needs to have a unique ID. This is the ID of this upload node, not the ID of the linked upload document */
-  id: string
-  relationTo: string
-  /** Value can be just the document ID, or the full, populated document */
-  value: number | string | TypedCollection
-}
+  [TCollectionSlug in CollectionSlug]: {
+    fields: TUploadExtraFieldsData
+    // Every lexical node that has sub-fields needs to have a unique ID. This is the ID of this upload node, not the ID of the linked upload document
+    id: string
+    relationTo: TCollectionSlug
+    // Value can be just the document ID, or the full, populated document
+    value: DataFromCollectionSlug<TCollectionSlug> | number | string
+  }
+}[CollectionSlug]
 
-// TODO: deprecate in Payload v4.
 /**
- * UploadDataImproved is a more precise type, and will replace UploadData in Payload v4.
- * This type is for internal use only as it will be deprecated in the future.
- * @internal
+ * @todo Replace UploadData with UploadDataImproved
  */
 export type UploadDataImproved<TUploadExtraFieldsData extends JsonObject = JsonObject> = {
-  fields: TUploadExtraFieldsData
-  // Every lexical node that has sub-fields needs to have a unique ID. This is the ID of this upload node, not the ID of the linked upload document
-  id: string
-  relationTo: string
-  // Value can be just the document ID, or the full, populated document
-  value: (FileData & TypeWithID) | number | string
-}
+  [TCollectionSlug in UploadCollectionSlug]: {
+    fields: TUploadExtraFieldsData
+    // Every lexical node that has sub-fields needs to have a unique ID. This is the ID of this upload node, not the ID of the linked upload document
+    id: string
+    relationTo: TCollectionSlug
+    // Value can be just the document ID, or the full, populated document
+    value: number | string | TypedUploadCollection[TCollectionSlug]
+  }
```

Comment:
I think `TypedUploadCollection[TCollectionSlug]` is significantly less precise than `(FileData & TypeWithID)`.

Before this PR:
![image](https://github.com/user-attachments/assets/69200f26-0dd6-42e0-b4ad-11c9abfe4215)

After this PR:
![image](https://github.com/user-attachments/assets/e0207c06-35b0-4878-a409-16b43acecc76)

Reply:
This will only be the case in our monorepo though - it will be more precise in a real project